### DISCUSSION TITLE
Prisma mock specific for single test file, or test cases.

### DISCUSSION THREAD

**Discussion by atahanozbayram** [prisma/prisma#13664] [Q&A]
Hello, I would like to learn a way to mock prisma client without using the ways used in the guide. My problem is that I want to mock prisma just for specific test files, or sometimes just for specific tests, and sometimes just use its real implementation therefore crud into some test database. I am using TypeScript, jest, and also express. Problem starts when I try to use singleton.ts code inside some of my test files. Then I run ts-jest to run the tests, but when I do that the special variable name for mockDeep thing is in a faulty behavior because the compiler changes mockDeep's name to jest_mock_extended_1.mockDeep or something like that, therefore mockDeep loses its privilege to being accessible before initialization. I need to get over this problem, but it has been for too long that I haven't figure it out. I will provide my repo for the project I have been trying to do this, but I don't think it will be helpful anyways.

Thanks for anyone that can solve my problem and whoever tries to do so.

Repo link: [https://github.com/atahanozbayram/whistler2](url)

**Comment by austincrim** [comment]
Hey @atahanozbayram 👋,

If you are using the Singleton method outlined in our docs, the solution would be to import and use the regular Prisma Client in the files you want the real version and import and use the mocked version from `singleton.ts` in the tests you want it to be mocked.

Does that solution fit your needs?

**Reply by atahanozbayram** [reply]
I have imported the prisma-client in the functions and stuff, and I did import the prisma-mock in the test files, but there are times that I need to just use real implementation and I don't want to use mockResolvedValue because I need to see what actual implementation of my code will end up with. But it seems I can't do that with setupAfterEnv thing having to mock prisma in every single test file I have.

**Reply by austincrim** [reply]
When you need the real implementation, import and instantiate the regular Prisma Client.

```ts
import { PrismaClient } from '@prisma/client'
const realPrismaClient = new PrismaClient()
```

**Reply by atahanozbayram** [reply]
Thanks for the reply, unfortunately your answer doesn't fit my problem.

I want to be able to mock prisma client per file like so if I don't mock it it should have the real implementation, or mock the prisma via setupFilesAfterEnv and then use real implementation in some files I do decide. What you suggest prevents me from having to use mocked  version in some test file and using real implementation in some file.  The code I have been trying to use but now works is as follows.




```ts
import { jest, describe, test } from "@jest/globals";
import { prisma } from "@shared/prisma-client";
import { PrismaClient, user } from "@prisma/client";
import { DeepMockProxy, mockDeep } from "jest-mock-extended";
import { saveUser } from "@root/src/shared/user";

jest.mock("@shared/prisma-client", () => {
	return {
		__esModule: true,
		prisma: mockDeep<PrismaClient>(),
	};
});

const prismaMock = prisma as unknown as DeepMockProxy<PrismaClient>;

describe("Finding ways for prisma mocking", () => {
	test("mocking without jest config", (done) => {
		const userInfo = {
			firstname: "Atahan",
			lastname: "Ozbayram",
			password: "Password1!",
			username: "username1",
			gender: 2,
			email: "atahan_ozbayram@hotmail.com",
			birth_date: new Date(1999, 7, 20),
		};

		prismaMock.user.create.mockResolvedValue(userInfo as unknown as user);

		saveUser(userInfo)
			.then(() => {
				expect(1).toBe(1);
				done();
			})
			.catch((error) => done(error));
	});
});
```

And I get some error like this. 
![image](https://user-images.githubusercontent.com/25008274/173617048-fbc28998-5b9e-434f-b700-bd59f6298877.png)


> When you need the real implementation, import and instantiate the regular Prisma Client.
> 
> ```ts
> import { PrismaClient } from '@prisma/client'
> const realPrismaClient = new PrismaClient()
> ```

**Reply by austincrim** [reply]
When you use `prismaMock`, that will be the mocked implementation, so if you need the real client, import and instantiated a regular `PrismaClient`, like I outlined above.

You can do this on a per-test basis since you will have separate instances of the client, one mocked and one real.

Unless I am missing something about your context.

**Reply by atahanozbayram** [reply]
I always use prisma in implementation code that in tutorial it is the client.ts and prisma-client.ts in my project, and I use singleton.ts which in my project is prisma-mock.ts in the tests. I don't use prismaMock in implementation, and even when I don't import it in some test files it is still mocking the tests because of setupFilesAfterEnv thing does run the singleton code and mocks the prisma whatever I do.

**Reply by austincrim** [reply]
Ah, I missed one thing. Try calling `jest.resetAllMocks()` in an `afterEach()` block. This should prevent the mocked Prisma Client from being persisted throughout all tests.

**Reply by atahanozbayram** [reply]
didn't do anything.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
